pkg/kn/core/root.go

```diff
@@ -212,6 +213,16 @@ func initConfig() {
 	}
 }
 
+func initConfigFlags() {
+	if commands.Cfg.PluginsDir == "" {
+		commands.Cfg.PluginsDir = viper.GetString("plugins-dir")
+	}
+
+	if commands.Cfg.LookupPlugins != viper.GetBool("lookup-plugins") {
+		commands.Cfg.LookupPlugins = viper.GetBool("lookup-plugins")
+	}
```

Comment:
You can as well remove the conditional, because when config == option stting it will be a no-op. In any case, you will override the configuration.

E.g. even when `--lookup-plugins` is not-set it defaults to `false`. That means that a `cfg.LookupPlugins == true` will be overwritten with `false` even when no `--lookup-plugins` is  given. I don't think that is what you want. For this to work, the flag needs to be a pointer with a `nil` value when not set.

Reply:
Yeah, I thought of that but not using a pointer. Perhaps makes sense. Let me see.

For tests, I am adding comprehensive tests in #308 since it's not so easy to do this test without more setup work to get config and load and call `kn`.

So if it's OK please approve this (after I change to pointer) so I can get the tests for #308 done.

Reply:
Actually @rhuss there is really no way to do this nicely for boolean flags. I changed it to pointer to a bool and we essentially end up with same code. 

The reason above works is that no `lookup-plugins` in the config will result in viper returning false which is the default value. And since this code will execute `PreRun` it will set the value to true if viper has that value. Naturally if the user passes a flag then that will override the flag.

I suggest we leave it as is. If you want to see the pointer code I can push it but it changes lots of files and ends up with the `initConfigFlags` as below:

```go
func initConfigFlags() {
	if commands.Cfg.PluginsDir == "" {
		commands.Cfg.PluginsDir = viper.GetString("plugins-dir")
	}

	var aBool bool
	// This condition if (...) never actually executes since setting 
	// commands.Cfg.LookupPlugins to nil ends up with SEG fault in adding the flag
	// to the root command so it has to be set to a pointer to a false bool
	if commands.Cfg.LookupPlugins == nil {
		aBool = viper.GetBool("lookup-plugins")
		commands.Cfg.LookupPlugins = &aBool
	} else {
		if viper.GetBool("lookup-plugins") {
			aBool = viper.GetBool("lookup-plugins")
			commands.Cfg.LookupPlugins = &aBool	
		} else {
			aBool = false
			commands.Cfg.LookupPlugins = &aBool
		}
	}
}
```

And that's above the ugly bool code we have to do since golang won't allow `&false` or `&true`.

Reply:
And of course as I mentioned, integration tests for #308 will cover this in details.

I changed the description and title as @sixolet mentions so I would say this is good to go IMO.

Reply:
> For this to work, the flag needs to be a pointer with a `nil` value when not set.

Not possible since that will cause a segmentation fault when adding the flag to the `rootCmd` in `plugins.go` function:

```go
func AddPluginFlags(cmd *cobra.Command) {
	cmd.Flags().StringVar(&commands.Cfg.PluginsDir, "plugins-dir", "~/.kn/plugins", "kn plugins directory")
	cmd.Flags().BoolVar(commands.Cfg.LookupPlugins, "lookup-plugins", false, "look for kn plugins in $PATH")
}
```

Reply:
> The reason above works is that no `lookup-plugins` in the config will result in viper returning false which is the default value. And since this code will execute `PreRun` it will set the value to true if viper has that value. Naturally if the user passes a flag then that will override the flag.

Sorry, I'm still not buying this. Let's assume that you don't set a configuration value for `lookup-plugins` but you set `--lookup-plugins` to true from the command line. What will `commands.Cfg.LookupPlugins` end up with ? Where have you seen that the values from command lines flags are injected *after* `PreRun` ?

With _using pointers_ I meant to use `cmd.Flags.BoolVarP()` variant to point to the address of `cfg.LookupPlugins`, not to change `cfg.LookupPlugins` to a pointer.

But again, you should bind the viper config values directly to the configuration as described in https://github.com/spf13/cobra#bind-flags-with-config and https://github.com/spf13/viper#working-with-flags . There is no need to deal with this low level stuff on our own.

Reply:
Hmm, maybe I really don't understand the issue at all, as we are using allready `BindPluginsFlagToViper()` for this kind of binding, so I'm not sure why we need a special treatment (and what the bug is actually).  

Could be have an issue with the binding because we introduced an own reconciliation ? And if we do an own handling, do we still need to call to `BindPluginsFlagForViper()` ?

Reply:
The binding is sort of one way Viper to file and file to Viper. But AFAIK does not load from Viper to variable. This is what I am doing in my change. Try it and see for yourself. The issue #467 shows how to repro.

Reply:
Can we solve this by conditioning on flags.IsSet()?

Reply:
Let me look into it @sixolet, interesting idea. Give me a day or so to investigate, trying to complete some things before Thxgiving (if possible).

Reply:
@sixolet the logic is simplified a bit but the result is the same. We still need to implicitly init the values from viper.

This works. I suggest we merge this in and move on. In future if someone wants to dig into how to save 10 lines of code via some Viper magic I am missing, I'd be happy to see it.